File: leetcode/medium/22. Generate Parentheses/solution.js
// https://leetcode.com/problems/generate-parentheses/
/**
 * @param {number} n
 * @return {string[]}
 */
function generateParenthesis(n) {
    const parens = [];
    const genParens = (open, close, len, str) => {
        if (str.length === len * 2) {
            parens.push(str);
            return;
        }

        if (open < len) genParens(open + 1, close, len, str + '(');
        if (close < open) genParens(open, close + 1, len, str + ')');
    };

    genParens(0, 0, n, '');
    return parens;
};

console.log(generateParenthesis(3));

File: leetcode/medium/22. Generate Parentheses/solution.cc
#include <vector>
#include <string>

using namespace std;

class Solution {
public:
    vector<string> generateParenthesis(int n) {
        vector<string> parens;
        gen_parens(parens, 0, 0, n, "");
        return parens;
    }

    void gen_parens(vector<string>& parens, int open, int close, int len, string str) {
        if (str.size() == 2 * len) {
            parens.push_back(str);
            return;
        }

        if (open < len) gen_parens(parens, open + 1, close, len, str + "(");
        if (close < open) gen_parens(parens, open, close + 1, len, str + ")");
    }
};